[Custom Device]第三方插件DenseTensor头文件访问受限

### bug描述 Describe the Bug

https://github.com/PaddlePaddle/Paddle/pull/51149
这个PR引起了Custom Device第三方插件的编译问题
[paddle/phi/core/selected_rows.h](https://github.com/PaddlePaddle/Paddle/pull/51149/files#diff-26278efd91168c9d5993373f5e80a5046a9f9e6bf94d3345facaa5a59e61ae2c)
set_type 和 set_layout是Custom Device下不可见的函数

这个由PADDLE_WITH_CUSTOM_KERNEL 这个宏引起的

https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/phi/core/dense_tensor.h
LINE 286
```
#ifndef PADDLE_WITH_CUSTOM_KERNEL
#include "paddle/phi/core/dense_tensor.inl"
#endif
```
PADDLE_WITH_CUSTOM_KERNEL 开启后看不到
https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/phi/core/dense_tensor.inl中的方法

如果一劳永逸地解决这些问题
是否能解除宏，让https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/phi/core/dense_tensor.inl对CustomDevice可见
```
// #ifndef PADDLE_WITH_CUSTOM_KERNEL
#include "paddle/phi/core/dense_tensor.inl"
// #endif
```



### 其他补充信息 Additional Supplementary Information

_No response_

编译报错
```
[build] In file included from ../third_party/paddle/include/paddle/phi/core/selected_rows.h:24,
[build]                  from ../third_party/paddle/include/paddle/phi/core/tensor_utils.h:19,
[build]                  from ../third_party/paddle/include/paddle/phi/core/kernel_context.h:24,
[build]                  from ../third_party/paddle/include/paddle/phi/include/core.h:20,
[build]                  from ../third_party/paddle/include/paddle/phi/extension.h:9,
[build]                  from ../source/paddle_opencl_plugin/xdx_conv2d_kernel.cc:3:
[build] ../third_party/paddle/include/paddle/phi/core/selected_rows_impl.h: In member function ‘void phi::SelectedRowsImpl::set_type(phi::DataType)’:
[build] ../third_party/paddle/include/paddle/phi/core/selected_rows_impl.h:162:49: error: ‘class phi::DenseTensor’ has no member named ‘set_type’
[build]   162 |   void set_type(const DataType dtype) { value_->set_type(dtype); }
[build]       |                                                 ^~~~~~~~
[build] ../third_party/paddle/include/paddle/phi/core/selected_rows_impl.h: In member function ‘void phi::SelectedRowsImpl::set_layout(phi::DataLayout)’:
[build] ../third_party/paddle/include/paddle/phi/core/selected_rows_impl.h:168:54: error: ‘class phi::DenseTensor’ has no member named ‘set_layout’; did you mean ‘layout’?
```
@YuanRisheng

--- a/paddle/phi/core/selected_rows.cc
+++ b/paddle/phi/core/selected_rows.cc
@@ -23,4 +23,10 @@ SelectedRows::SelectedRows(const std::vector<int64_t>& rows,
 SelectedRows::SelectedRows()
     : impl_(std::make_shared<phi::SelectedRowsImpl>()) {}
 
+void SelectedRows::set_type(const DataType dtype) { impl_->set_type(dtype); }
+
+void SelectedRows::set_layout(const DataLayout layout) {
+  impl_->set_layout(layout);
+}
+
 }  // namespace phi

--- a/paddle/phi/core/selected_rows.h
+++ b/paddle/phi/core/selected_rows.h
@@ -139,13 +139,17 @@ class SelectedRows : public TensorBase,
   /// \return The data type of the tensor.
   DataType dtype() const noexcept override { return impl_->dtype(); }
 
-  void set_type(const DataType dtype) { impl_->set_type(dtype); }
+#ifndef PADDLE_WITH_CUSTOM_KERNEL
+  void set_type(const DataType dtype);
+#endif
 
   /// \brief Returns the data layout of the tensor.
   /// \return The data layout of the tensor.
   DataLayout layout() const noexcept override { return impl_->layout(); }
 
-  void set_layout(const DataLayout layout) { impl_->set_layout(layout); }
+#ifndef PADDLE_WITH_CUSTOM_KERNEL
+  void set_layout(const DataLayout layout);
+#endif
 
   /// \brief Returns the data place of the tensor.
   /// \return The data place of the tensor.

--- a/paddle/phi/core/selected_rows_impl.cc
+++ b/paddle/phi/core/selected_rows_impl.cc
@@ -208,4 +208,13 @@ void SelectedRowsImpl::Get(const phi::DenseTensor& ids,
     }
   }
 }
+
+void SelectedRowsImpl::set_type(const DataType dtype) {
+  value_->set_type(dtype);
+}
+
+void SelectedRowsImpl::set_layout(const DataLayout layout) {
+  value_->set_layout(layout);
+}
+
 }  // namespace phi

--- a/paddle/phi/core/selected_rows_impl.h
+++ b/paddle/phi/core/selected_rows_impl.h
@@ -159,13 +159,17 @@ class SelectedRowsImpl {
   /// \return The data type of the tensor.
   DataType dtype() const noexcept { return value_->dtype(); }
 
-  void set_type(const DataType dtype) { value_->set_type(dtype); }
+#ifndef PADDLE_WITH_CUSTOM_KERNEL
+  void set_type(const DataType dtype);
+#endif
 
   /// \brief Returns the data layout of the tensor.
   /// \return The data layout of the tensor.
   DataLayout layout() const noexcept { return value_->layout(); }
 
-  void set_layout(const DataLayout layout) { value_->set_layout(layout); }
+#ifndef PADDLE_WITH_CUSTOM_KERNEL
+  void set_layout(const DataLayout layout);
+#endif
 
   /// \brief Returns the data place of the tensor.
   /// \return The data place of the tensor.

--- a/paddle/phi/core/sparse_coo_tensor.cc
+++ b/paddle/phi/core/sparse_coo_tensor.cc
@@ -84,6 +84,12 @@ int64_t SparseCooTensor::nnz() const {
   }
 }
 
+void SparseCooTensor::set_type(const DataType dtype) { meta_.dtype = dtype; }
+
+void SparseCooTensor::set_layout(const DataLayout layout) {
+  meta_.layout = layout;
+}
+
 void SparseCooTensor::Resize(const DDim& dense_dims,
                              const int64_t sparse_dim,
                              const int64_t non_zero_num) {

--- a/paddle/phi/core/sparse_coo_tensor.h
+++ b/paddle/phi/core/sparse_coo_tensor.h
@@ -104,13 +104,18 @@ class SparseCooTensor : public TensorBase,
   /// \brief Returns the data type of the tensor.
   /// \return The data type of the tensor.
   DataType dtype() const noexcept override { return meta_.dtype; }
-  void set_type(const DataType dtype) { meta_.dtype = dtype; }
+
+#ifndef PADDLE_WITH_CUSTOM_KERNEL
+  void set_type(const DataType dtype);
+#endif
 
   /// \brief Returns the data layout of the tensor.
   /// \return The data layout of the tensor.
   DataLayout layout() const noexcept override { return meta_.layout; }
 
-  void set_layout(const DataLayout layout) { meta_.layout = layout; }
+#ifndef PADDLE_WITH_CUSTOM_KERNEL
+  void set_layout(const DataLayout layout);
+#endif
 
   /// \brief Returns the data place of the tensor.
   /// \return The data place of the tensor.

--- a/paddle/phi/core/sparse_csr_tensor.cc
+++ b/paddle/phi/core/sparse_csr_tensor.cc
@@ -88,6 +88,12 @@ void* SparseCsrTensor::AllocateFrom(Allocator* allocator,
       allocator, dtype, requested_size, fake_alloc);
 }
 
+void SparseCsrTensor::set_type(const DataType dtype) { meta_.dtype = dtype; }
+
+void SparseCsrTensor::set_layout(const DataLayout layout) {
+  meta_.layout = layout;
+}
+
 void SparseCsrTensor::Resize(const DDim& dense_dims,
                              const int64_t non_zero_num) {
   PADDLE_ENFORCE(this->initialized(),

--- a/paddle/phi/core/sparse_csr_tensor.h
+++ b/paddle/phi/core/sparse_csr_tensor.h
@@ -110,13 +110,17 @@ class SparseCsrTensor : public TensorBase,
   /// \return The data type of the tensor.
   DataType dtype() const noexcept override { return meta_.dtype; }
 
-  void set_type(const DataType dtype) { meta_.dtype = dtype; }
+#ifndef PADDLE_WITH_CUSTOM_KERNEL
+  void set_type(const DataType dtype);
+#endif
 
   /// \brief Returns the data layout of the tensor.
   /// \return The data layout of the tensor.
   DataLayout layout() const noexcept override { return meta_.layout; }
 
-  void set_layout(const DataLayout layout) { meta_.layout = layout; }
+#ifndef PADDLE_WITH_CUSTOM_KERNEL
+  void set_layout(const DataLayout layout);
+#endif
 
   /// \brief Returns the data place of the tensor.
   /// \return The data place of the tensor.

--- a/paddle/phi/core/tensor_array.h
+++ b/paddle/phi/core/tensor_array.h
@@ -65,11 +65,15 @@ class TensorArray : public TensorBase,
 
   DataType dtype() const override;
 
+#ifndef PADDLE_WITH_CUSTOM_KERNEL
   void set_type(const DataType dtype);
+#endif
 
   DataLayout layout() const override;
 
+#ifndef PADDLE_WITH_CUSTOM_KERNEL
   void set_layout(const DataLayout layout);
+#endif
 
   /// \brief This overrided function is not used in TensorArray.
   bool valid() const override;

